packages/terra-time-input/src/TimeInput.jsx

```diff
@@ -146,32 +263,171 @@ class TimeInput extends React.Component {
-    if (type === inputType.HOUR && inputValue.length === 2) {
+    if (inputValue.length === 2) {
       this.minuteInput.textInput.focus();
     }
 
-    this.handleValueChange(event, type, inputValue);
+    this.handleValueChange(event, inputType.HOUR, inputValue, this.state.meridiem);
   }
 
-  handleHourInputKeyDown(event) {
-    this.handleInputKeyDown(event, inputType.HOUR);
+  handleMinuteChange(event) {
+    // Ignore the input and don't update the state if the entered value is a non numeric character.
+    if (event.target.value.length > 0) {
+      const isNumeric = /^\d+$/.test(event.target.value);
+
+      if (!isNumeric) {
+        return;
+      }
+    }
+
+    let inputValue = event.target.value;
+    const stateValue = this.state.minute;
+    const maxValue = 59;
+
+    // Ignore the entry if the value did not change or it is invalid.
+    // When 'Predictive text' is enabled on Android the maxLength attribute on the input is ignored so we have to the length of inputValue to make sure that it is less then 2.
+    if (inputValue === stateValue || inputValue.length > 2 || Number(inputValue) > maxValue) {
+      return;
+    }
+
+    // If the change made was not a deletion of a digit,
+    // then prepend '0' if the input value is a single digit value between 3 and 9 for hour or 6 and 9 for minute.
+    if (inputValue.length >= stateValue.length) {
+      const digitsToPrependZero = ['6', '7', '8', '9'];
+      if (digitsToPrependZero.indexOf(inputValue) > -1) {
+        inputValue = '0'.concat(inputValue);
+      }
+    }
+
+    // Move focus to the merdiem for 12 hours times if the minute input has a valid and complete entry.
+    if (this.props.variant === FORMAT_12_HOUR && inputValue.length === 2 && this.meridiemInput) {
+      this.meridiemInput.focus();
+    }
+
+    this.handleValueChange(event, inputType.MINUTE, inputValue, this.state.meridiem);
   }
 
-  handleMinuteInputKeyDown(event) {
-    this.handleInputKeyDown(event, inputType.MINUTE);
+  handleMeridiemChange(event) {
+    this.setState({
+      meridiem: event.target.value,
+    });
+
+    this.handleValueChange(event, inputType.HOUR, this.state.hour.toString(), event.target.value);
   }
 
-  handleInputKeyDown(event, type) {
-    let stateValue = type === inputType.HOUR ? this.state.hour : this.state.minute;
-    const maxValue = type === inputType.HOUR ? 23 : 59;
+  meridiemFocus() {
+    this.setState({
+      isFocused: true,
+      meridiemFocused: true,
+    });
+
+    this.meridiemInput.focus();
+  }
+
+  meridiemBlur() {
+    this.setState({
+      isFocused: false,
+      meridiemFocused: false,
+    });
+  }
+
+  handleHourInputKeyDown(event) {
+    let stateValue = this.state.hour;
+    const maxValue = this.props.variant === FORMAT_12_HOUR ? 12 : 23;
+    const minValue = this.props.variant === FORMAT_12_HOUR ? 1 : 0;
+
+    let meridiem = this.state.meridiem;
+    let updateStateValue = false;
+
+    if (event.keyCode === keyCodes.ARROWUP) {
```

Comment:
this method is quite large, can we break this up
I tend to go with the rule that it there is over 10 ways to get through a method it should be broken apart

maybe could put some of the minute calculation in a private method

Reply:
I agree, breaking these into a smaller function would make this more readable. Possibly add a func to TimeUtils that will break each if into its own sub-function with descriptive names maybe?

Reply:
I can't move them all to the timeUtils as some deal with individual attributes attached to this react class, but I should be able to create a functions that properly increment and decrement the minutes

Reply:
Try this refactor approach. https://github.com/cerner/terra-core/pull/1027/commits/45b44e889215a8dec2e1aed334cdb59ae3670902